src/privatesend/privatesend.cpp

```diff
@@ -224,6 +224,102 @@ std::string CPrivateSendBaseSession::GetStateString() const
+{
+    std::set<CScript> setScripPubKeys;
+    nMessageIDRet = MSG_NOERR;
+    if (fConsumeCollateralRet) *fConsumeCollateralRet = false;
+
+    if (vin.size() != vout.size()) {
+        LogPrint(BCLog::PRIVATESEND, "CPrivateSendBaseSession::%s -- ERROR: inputs vs outputs size mismatch! %d vs %d\n", __func__, vin.size(), vout.size());
+        nMessageIDRet = ERR_SIZE_MISMATCH;
+        if (fConsumeCollateralRet) *fConsumeCollateralRet = true;
+        return false;
+    }
+
+    auto checkTxOut = [&](const CTxOut& txout) {
+        std::vector<CTxOut> vecTxOut{txout};
+        int nDenom = CPrivateSend::GetDenominations(vecTxOut);
+        if (nDenom != nSessionDenom) {
+            LogPrint(BCLog::PRIVATESEND, "CPrivateSendBaseSession::IsValidInOuts -- ERROR: incompatible denom %d (%s) != nSessionDenom %d (%s)\n",
+                    nDenom, CPrivateSend::GetDenominationsToString(nDenom), nSessionDenom, CPrivateSend::GetDenominationsToString(nSessionDenom));
+            nMessageIDRet = ERR_DENOM;
+            if (fConsumeCollateralRet) *fConsumeCollateralRet = true;
+            return false;
+        }
+        if (!txout.scriptPubKey.IsPayToPublicKeyHash()) {
+            LogPrint(BCLog::PRIVATESEND, "CPrivateSendBaseSession::IsValidInOuts -- ERROR: invalid script! scriptPubKey=%s\n", ScriptToAsmStr(txout.scriptPubKey));
+            nMessageIDRet = ERR_INVALID_SCRIPT;
+            if (fConsumeCollateralRet) *fConsumeCollateralRet = true;
+            return false;
+        }
+        if (!setScripPubKeys.insert(txout.scriptPubKey).second) {
+            LogPrint(BCLog::PRIVATESEND, "CPrivateSendBaseSession::IsValidInOuts -- ERROR: already have this script! scriptPubKey=%s\n", ScriptToAsmStr(txout.scriptPubKey));
+            nMessageIDRet = ERR_ALREADY_HAVE;
+            if (fConsumeCollateralRet) *fConsumeCollateralRet = true;
+            return false;
+        }
+        // IsPayToPublicKeyHash() above already checks for scriptPubKey size,
+        // no need to double check, hence no usage of ERR_NON_STANDARD_PUBKEY
+        return true;
+    };
+
+    CAmount nFees{0};
+
+    for (const auto& txout : vout) {
+        if (!checkTxOut(txout)) {
+            return false;
+        }
+        nFees -= txout.nValue;
+    }
+
+    for (const auto& txin : vin) {
+        LogPrint(BCLog::PRIVATESEND, "CPrivateSendBaseSession::%s -- txin=%s\n", __func__, txin.ToString());
+
+        if (txin.prevout.IsNull()) {
+            LogPrint(BCLog::PRIVATESEND, "CPrivateSendBaseSession::%s -- ERROR: invalid input!\n", __func__);
+            nMessageIDRet = ERR_INVALID_INPUT;
+            if (fConsumeCollateralRet) *fConsumeCollateralRet = true;
+            return false;
+        }
+
+        Coin coin;
+        CScript scriptPubKeyIn;
+        auto mempoolTx = mempool.get(txin.prevout.hash);
```

Comment:
btw, for this type of logic (check mempool and then UTXO set) we have the `CCoinsViewMemPool`, which does exactly this. See 2c00a2c851638635b3e0645aef8ae4c7ca88230f